### DISCUSSION TITLE
Type narrowing with missing generic argument

### DISCUSSION THREAD

**Discussion by cdce8p** [python/typing#1108] [Q&A]
Came across something like this today
```py
from typing import Generic, TypeVar

_T = TypeVar("_T", bound="DataUpdateCoordinator")

class DataUpdateCoordinator:
    data: str | int

class CoordinatorEntity(Generic[_T]):
    def __init__(self, coordinator: _T) -> None:
        self.coordinator = coordinator

class A(CoordinatorEntity):
    def __init__(self) -> None:
        reveal_type(self.coordinator)
        reveal_type(self.coordinator.data)
        print(self.coordinator.error)  # This should be an error
```

`CoordinatorEntity` for class `A` is (obviously) missing a generic argument. Let's assume for a moment that it's not possible to add one in a legacy codebase. Can `self.coordinator` still be narrowed to `DataUpdateCoordinator`?

mypy and pyre just infer it as `Any` whereas pyright says it's `unknown`. We do know that `_T` has an upper bound in `DataUpdateCoordinator` so every generic argument will need to be a subclass of `DataUpdateCoordinator` and thus also have a `data` attribute. `error` is more complicate. Without a concrete type, I would assume this should be an error though.

Am I missing something? If not, improving the type narrowing here could help with better intellisense suggestions and type checking in partially typed code bases.

**Comment by erictraut** [comment]
The type of `coordinator` is `_T@CoordinatorEntry` (i.e. `_T` scoped to the class `CoordinatorEntry`). Outside of the class, its type is dictated by the specialization of `CoordinatorEntity`. In your particular example, `CoordinatorEntity` has been specialized implicitly with `Any` because of the missing type argument. That means the type of `self.coordinator` is `Any`. So the behavior of all three type checkers is correct in this case.
